### ISSUE TITLE
False positive in NonStaticMethodCheck

### ISSUE THREAD

**Issue by mkordas** [yegor256/qulice#806]
On the following code Qulice complains with `This method must be static, because it does not refer to "this" (NonStaticMethodCheck)`

``` java
class WithInner {
    public Inner inner() {
        return new Inner();
    }
    class Inner {
    }
}
```

However, I cannot make this method static, as `new Inner()` cannot be referenced from static context.

Please disable `NonStaticMethodCheck` for all methods accessing inner classes.

**Comment by krzyk** [comment]
@davvd valid bug

**Comment by davvd** [comment]
> @davvd valid bug

@krzyk tagged this issue with "bug"

**Comment by davvd** [comment]
@mkordas thanks for reporting! I topped your account for 15 mins, transaction 57c42a636d5f4b040b19fa15

**Comment by amihaiemil** [comment]
@mkordas @krzyk Also this rule could be taken out entirely? It creates a vicious circle together with "no public static methods" rule. Check out [this](https://github.com/yegor256/netbout/blob/master/netbout-client/src/test/java/com/netbout/client/NbRule.java) class: method get() is currently ``public static``, and there is that ``todo`` there. With qulice 0.17.1, which contains this rule, you have to following: 

1) make ``get()`` method default static, so you won't have public static methods. Then qulice complains that you should **explicetly specify a modifier.** You specify ``protected``, as it cannot be private, and then it complains that **final classes cannot have protected methods**

2) make ``get()`` method public and not static, you get **This method must be static, because it does not refer to "this" (NonStaticMethodCheck)**

In fact, the problem is that the "no public static methods" rule and NonStaticMethodCheck are against each other.

**Comment by mkordas** [comment]
@amihaiemil but the class from your example is not a proper object, it needs to be redesigned entirely. See http://www.yegor256.com/2014/11/20/seven-virtues-of-good-object.html#5-his-class-doesn-39-t-have-anything-static

**Comment by mkordas** [comment]
@amihaiemil 
> In fact, the problem is that the "no public static methods" rule and NonStaticMethodCheck are against each other.

Can you elaborate? They are just against utility classes: http://www.yegor256.com/2014/05/05/oop-alternative-to-utility-classes.html

**Comment by amihaiemil** [comment]
@mkordas I have already elaborated. 

The problem that needs to be fixed in the class is the public static method (as the todo states). In my first comment, points 1) and 2) are the options to fix this problem, and none of them works fine. 1 colides with some qulice rules, while 2 colides with the NonStaticMethodCheck

See the problem now?

**Comment by mkordas** [comment]
@amihaiemil but you need to choose third option, which is to redesign a class to be in-line with OOP. Right now it doesn't contain any field and doesn't encapsulate any state, so this needs to be changed.

**Comment by yegor256** [comment]
@mkordas this is fixed in the latest version (0.23.0)
